compaction/compaction.cc

```diff
@@ -1703,6 +1704,98 @@ future<uint64_t> scrub_validate_mode_validate_reader(flat_mutation_reader_v2 rea
+    const auto& schema = *sst->get_schema();
+    partition_key::tri_compare cmp(schema);
+    std::optional<partition_key> last_pkey;
+    std::optional<uint64_t> last_pos;
+
+    std::exception_ptr ex;
+    try {
+        while (!idx_reader.eof()) {
+            if (cdata.is_stop_requested()) [[unlikely]] {
+                // Compaction manager will catch this exception and re-schedule the compaction.
+                throw compaction_stopped_exception(schema.ks_name(), schema.cf_name(), cdata.stop_requested);
+            }
+
+            co_await idx_reader.read_partition_data();
+            auto pos = idx_reader.get_data_file_position();
+            auto pkey = idx_reader.get_partition_key();
+
+            if (last_pkey && cmp(*last_pkey, pkey) <= 0) {
+                ++errors;
+                clogger.error("[scrub compaction {}.{}] Out-of-order partition {} ({}) in index, previous partition is {} ({})",
+                        schema.ks_name(),
+                        schema.cf_name(),
+                        scrub_compaction::partition_key_to_string(pkey, schema),
+                        pkey,
+                        scrub_compaction::partition_key_to_string(*last_pkey, schema),
+                        *last_pkey);
+            }
+
+            if (last_pos && pos <= *last_pos) {
+                ++errors;
+                clogger.error("[scrub compaction {}.{}] Out-of-order position {} for partition {} ({}) in index, previous position is {} for partition {} ({})",
+                        schema.ks_name(),
+                        schema.cf_name(),
+                        pos,
+                        scrub_compaction::partition_key_to_string(pkey, schema),
+                        pkey,
+                        *last_pos,
+                        scrub_compaction::partition_key_to_string(*last_pkey, schema),
+                        *last_pkey);
+            }
+
+            auto pr = dht::partition_range::make_singular(dht::decorate_key(schema, pkey));
+            auto reader = sst->make_reader(sst->get_schema(), permit, pr, schema.full_slice(), pc, {}, streamed_mutation::forwarding::no, mutation_reader::forwarding::no);
+            try {
+                auto maybe_ps_opt = co_await reader();
+                if (!maybe_ps_opt) {
+                    ++errors;
+                    clogger.error("[scrub compaction {}.{}] Invalid index entry for partition {} ({}), partition not found in data file",
+                            schema.ks_name(), schema.cf_name(), scrub_compaction::partition_key_to_string(pkey, schema), pkey);
+                } else if (!maybe_ps_opt->is_partition_start()) {
+                    ++errors;
+                    clogger.error("[scrub compaction {}.{}] Invalid index entry for partition {} ({}), entry doesn't point to start of partition",
+                            schema.ks_name(), schema.cf_name(), scrub_compaction::partition_key_to_string(pkey, schema), pkey);
+                } else {
```

Comment:
Should check that the the partition key matches. But not if the partition key comes from the input to make_reader (rather than being read from the sstable). I think it's bad to rely on single-partition reads.